modules/imgproc/src/peronamalik.cpp

```diff
@@ -0,0 +1,89 @@
+
+#include "precomp.hpp"
+#include "opencl_kernels_imgproc.hpp"
+#include "cvconfig.h"
+
+#include <iostream>
+using namespace std;
+
+namespace cv
+{
+    void PeronaMalik(InputArray src, OutputArray dst, double timeStepSize, double k, int noOfTimeSteps, int fluxFunc )
+    {
+        CV_Assert(k != 0);
+        CV_Assert(timeStepSize > 0);
+        CV_Assert(noOfTimeSteps >= 0);
+        CV_Assert((fluxFunc == CV_PERONA_MALIK_EXPONENTIAL) || (fluxFunc == CV_PERONA_MALIK_INVERSE_QUADRATIC));
+        CV_Assert(src.size() == dst.size());
+        CV_Assert(src.type() == dst.type());
```

Comment:
OpenCV functions are usually responsible for allocation of output images.

This algorithm can be implemented like this:
```
CV_Assert(src.type() == CV_64FC1);
Mat _src = src.getMat();
dst.createSameSize(src, src.type())
Mat _dst = dst.getMat();

if (
    noOfTimeSteps > 1 // _src is reused as intermediate buffer
    || _src.data == _dst.data // support inplace processing: PeronaMalik(img, img, ...)
)
{
    _src = _src.clone();
}

...
for (int t = 0; t < noOfTimeSteps; t++)
{
    ... update _dst based on _src ...

    swap(_src, _dst); // note: this doesn't change "src/dst" InputArrays
}

if (noOfTimeSteps % 2 == 0)
{
    _src.copyTo(dst); // note: changed _dst => dst
};
```